### PR TITLE
PIM-10212: Prevent spaces in locale codes

### PR DESCRIPTION
**Pull request by Balamung**
**Description (for Contributor and Core Developer)**

https://akeneo.atlassian.net/browse/PIM-10212

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [x] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Channel/Component/Validator/Constraint/LocaleCodeValidator.php [hunk 1]
```diff
@@ -9,7 +9,7 @@ final class LocaleCodeValidator extends ConstraintValidator
 {
     public function validate($value, Constraint $constraint): void
     {
-        if (! preg_match('#[a-z]{2,3}_[a-z0-9_]{2,}#i', $value)) {
+        if (! preg_match('#^[a-z0-9_]*[a-z]{2,3}_[a-z0-9_]{2,}[a-z0-9_]*$#i', $value)) {
```
Comment (pierallard) [inline src/Akeneo/Channel/Component/Validator/Constraint/LocaleCodeValidator.php:12]: [a-z0-9_]{2,}[a-z0-9_]*
is not equivalent to
[a-z0-9_]{2,}
?

File: upgrades/schema/Version_6_0_20211222080930_trim_locales_codes.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+declare(strict_types=1);
+
+namespace Pim\Upgrade\Schema;
+
+use Doctrine\DBAL\Schema\Schema;
+use Doctrine\Migrations\AbstractMigration;
+
+final class Version_6_0_20211222080930_trim_locales_codes extends AbstractMigration
+{
+    public function getDescription(): string
```
Comment (pierallard) [inline upgrades/schema/Version_6_0_20211222080930_trim_locales_codes.php:12]: I hink you can remove this method

File: upgrades/test_schema/Version_6_0_20211222080930_trim_locales_codes_Integration.php [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    private function getConnection(): Connection
+    {
+        return $this->get('database_connection');
+    }
+
+    private function addLocaleWithExtraSpaces(): void
+    {
+        $this->getConnection()->executeStatement(<<<SQL
+            INSERT INTO pim_catalog_locale (code, is_activated)
+            VALUES (' global_fr', 1);
```
Comment (pierallard) [inline upgrades/test_schema/Version_6_0_20211222080930_trim_locales_codes_Integration.php:40]: What if the locale was "-fr_FR" ?

File: src/Akeneo/Channel/Component/Validator/Constraint/LocaleCodeValidator.php (hunk 1 above)
Comment (Balamung) [reply src/Akeneo/Channel/Component/Validator/Constraint/LocaleCodeValidator.php:12]: Yes ! It goes with your 3rd comment, should I still allow dashes in the 2 parts I added ? I'm not sure cause it seems bad to leave, but 2 clients have some locales with dashes according to big query.

File: src/Akeneo/Channel/Component/Validator/Constraint/LocaleCodeValidator.php [hunk 4]
```diff
@@ -9,6 +9,15 @@ final class LocaleCodeValidator extends ConstraintValidator
...
     public function validate($value, Constraint $constraint): void
     {
+        // PIM-10212 For newly created locales, we do a proper check, for existing ones, we keep the old check to avoid BC breaks
+        $locale = $this->context->getRoot();
+        if (null !== $locale && null === $locale->getId()) {
+            if (! preg_match('#^[a-z0-9]{2,}_[a-z0-9_]{2,}$#i', $value)) {
+                $this->context->buildViolation($constraint->message)->addViolation();
+            }
+            return;
+        }
```
Comment (mmetayer) [inline src/Akeneo/Channel/Component/Validator/Constraint/LocaleCodeValidator.php:19]:
```suggestion
        if (!$locale instanceof LocaleInterface) {
            return;
        }
        if (null === $locale->getId() && !\preg_match('#^[a-z]{2,}_[a-z0-9_]{2,}$#i', $value)) {
                $this->context->buildViolation($constraint->message)->addViolation();
                
                return;
         }
```

File: tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php [hunk 5]
```diff
@@ -29,19 +33,57 @@ function it_validates_a_valid_locale_code($context) {
...
+        $context->getRoot()->willReturn($contextLocale);
+        $contextLocale->getId()->willReturn('fr_FR');
+
+        $violation->addViolation()->shouldNotBeCalled();
+
+        $this->validate('aa_AA-test', $constraint);
+        $this->validate(' aa_AA', $constraint);
+    }
+
+    function it_add_a_violation_with_an_invalid_locale_code_with_existing_locale(ConstraintViolationBuilderInterface $violation, $context, Locale $contextLocale)
```
Comment (mmetayer) [inline tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php:71]:
```suggestion
    function it_adds_a_violation_with_an_invalid_locale_code_with_existing_locale(ConstraintViolationBuilderInterface $violation, $context, LocaleInterface $contextLocale)
```

File: tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php [hunk 6]
```diff
@@ -29,19 +33,57 @@ function it_validates_a_valid_locale_code($context) {
         $this->validate('en_029', $constraint);
     }
 
-    function it_add_a_violation_with_an_invalid_locale_code(ConstraintViolationBuilderInterface $violation, $context) {
+    function it_add_a_violation_with_an_invalid_locale_code(ConstraintViolationBuilderInterface $violation, $context, Locale $contextLocale)
```
Comment (mmetayer) [inline tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php:36]:
```suggestion
    function it_adds_a_violation_with_an_invalid_locale_code(ConstraintViolationBuilderInterface $violation, $context, Locale $contextLocale)
```

File: tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php [hunk 7]
```diff
@@ -15,10 +16,13 @@ function let(ExecutionContextInterface $context)
         $this->initialize($context);
     }
 
-    function it_validates_a_valid_locale_code($context) {
+    function it_validates_a_valid_locale_code($context, Locale $contextLocale)
```
Comment (mmetayer) [inline tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php:19]:
```suggestion
    function it_validates_a_valid_locale_code($context, LocaleInterface $contextLocale)
```

File: tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php (hunk 6 above)
Comment (mmetayer) [inline tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php:36]:
```suggestion
    function it_add_a_violation_with_an_invalid_locale_code(ConstraintViolationBuilderInterface $violation, $context, LocaleInterface $contextLocale)
```

File: tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php [hunk 8]
```diff
@@ -29,19 +33,57 @@ function it_validates_a_valid_locale_code($context) {
...
+        $this->validate('aA', $constraint);
+        $this->validate('AA_', $constraint);
+        $this->validate('aA_A', $constraint);
+        $this->validate('A_AA', $constraint);
+        $this->validate(' aa_AA', $constraint);
+        $this->validate('aa_AA-test ', $constraint);
+        $this->validate('aa AA ', $constraint);
+    }
+
+    function it_validates_an_old_pattern_if_locale_already_exists(ConstraintViolationBuilderInterface $violation, $context, Locale $contextLocale)
```
Comment (mmetayer) [inline tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php:57]:
```suggestion
    function it_validates_an_old_pattern_if_locale_already_exists(ConstraintViolationBuilderInterface $violation, $context, LocaleInterface $contextLocale)
```

File: tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php [hunk 9]
```diff
@@ -29,19 +33,57 @@ function it_validates_a_valid_locale_code($context) {
...
+        $this->validate('aa_AA-test', $constraint);
+        $this->validate(' aa_AA', $constraint);
+    }
+
+    function it_add_a_violation_with_an_invalid_locale_code_with_existing_locale(ConstraintViolationBuilderInterface $violation, $context, Locale $contextLocale)
+    {
         $constraint = new LocaleCode();
         $context->buildViolation($constraint->message)->willReturn($violation);
+        $context->getRoot()->willReturn($contextLocale);
+        $contextLocale->getId()->willReturn(null);
```
Comment (mmetayer) [inline tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php:76]: ?

**Review by nmarniesse — APPROVED** [review]

File: tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php [hunk 10]
```diff
@@ -27,21 +31,60 @@ function it_validates_a_valid_locale_code($context) {
...
+        $this->validate('aa_AA-test', $constraint);
+        $this->validate(' aa_AA', $constraint);
+    }
+
+    function it_adds_a_violation_with_an_invalid_locale_code_with_existing_locale(ConstraintViolationBuilderInterface $violation, $context, LocaleInterface $contextLocale)
+    {
+        $constraint = new LocaleCode();
+        $context->buildViolation($constraint->message)->willReturn($violation);
+        $context->getRoot()->willReturn($contextLocale);
+        $contextLocale->getId()->willReturn('fr_FR');
```
Comment (mmetayer) [inline tests/back/Channel/Specification/Component/Validator/Constraint/LocaleCodeValidatorSpec.php:79]:
```suggestion
        $contextLocale->getId()->willReturn(42);
```

**Review by mmetayer — APPROVED** [review]

**Review by jmleroux — APPROVED** [review]

**Pull request merged**